File: dp/c/tag_dispatch.c
// Tag Dispatch Pattern in C
// Select behavior at compile/link time using function pointers keyed by tag

#include <stdio.h>
#include <string.h>

typedef enum { FMT_JSON, FMT_XML, FMT_CSV } Format;

typedef struct { const char* key; const char* value; } KV;

void serialize_json(const KV* data, int n, char* out, int out_size) {
    int pos = 0;
    pos += snprintf(out + pos, out_size - pos, "{");
    for (int i = 0; i < n; i++) {
        if (i > 0) pos += snprintf(out + pos, out_size - pos, ",");
        pos += snprintf(out + pos, out_size - pos, "\"%s\":\"%s\"", data[i].key, data[i].value);
    }
    snprintf(out + pos, out_size - pos, "}");
}

void serialize_xml(const KV* data, int n, char* out, int out_size) {
    int pos = 0;
    pos += snprintf(out + pos, out_size - pos, "<record>");
    for (int i = 0; i < n; i++)
        pos += snprintf(out + pos, out_size - pos, "<%s>%s</%s>", data[i].key, data[i].value, data[i].key);
    snprintf(out + pos, out_size - pos, "</record>");
}

void serialize_csv(const KV* data, int n, char* out, int out_size) {
    int pos = 0;
    for (int i = 0; i < n; i++)
        pos += snprintf(out + pos, out_size - pos, "%s%s", i ? "," : "", data[i].key);
    pos += snprintf(out + pos, out_size - pos, "\n");
    for (int i = 0; i < n; i++)
        pos += snprintf(out + pos, out_size - pos, "%s%s", i ? "," : "", data[i].value);
}

typedef void (*Serializer)(const KV*, int, char*, int);

Serializer get_serializer(Format fmt) {
    switch (fmt) {
        case FMT_JSON: return serialize_json;
        case FMT_XML: return serialize_xml;
        case FMT_CSV: return serialize_csv;
    }
    return serialize_json;
}

int main(void) {
    KV data[] = {{"name", "Alice"}, {"age", "30"}, {"city", "Seoul"}};
    int n = 3;
    char buf[256];

    Format formats[] = {FMT_JSON, FMT_XML, FMT_CSV};
    const char* names[] = {"JSON", "XML", "CSV"};

    for (int i = 0; i < 3; i++) {
        Serializer s = get_serializer(formats[i]);
        s(data, n, buf, sizeof(buf));
        printf("%s: %s\n", names[i], buf);
    }
    return 0;
}

File: dp/rust/src/tag_dispatch.rs
// Tag Dispatch Pattern in Rust
// Use marker types (zero-sized) to select behavior at compile time

use std::marker::PhantomData;

// Tags (zero-sized types)
struct Json;
struct Xml;
struct Csv;

// Serializer parameterized by format tag
struct Serializer<Format> {
    _format: PhantomData<Format>,
}

trait Serialize {
    fn serialize(&self, data: &[(&str, &str)]) -> String;
}

impl Serialize for Serializer<Json> {
    fn serialize(&self, data: &[(&str, &str)]) -> String {
        let fields: Vec<String> = data
            .iter()
            .map(|(k, v)| format!("\"{}\":\"{}\"", k, v))
            .collect();
        format!("{{{}}}", fields.join(","))
    }
}

impl Serialize for Serializer<Xml> {
    fn serialize(&self, data: &[(&str, &str)]) -> String {
        let fields: Vec<String> = data
            .iter()
            .map(|(k, v)| format!("<{}>{}</{}>", k, v, k))
            .collect();
        format!("<record>{}</record>", fields.join(""))
    }
}

impl Serialize for Serializer<Csv> {
    fn serialize(&self, data: &[(&str, &str)]) -> String {
        let header: Vec<&str> = data.iter().map(|(k, _)| *k).collect();
        let values: Vec<&str> = data.iter().map(|(_, v)| *v).collect();
        format!("{}\n{}", header.join(","), values.join(","))
    }
}

fn make_serializer<F>() -> Serializer<F> {
    Serializer { _format: PhantomData }
}

// Tag dispatch for sorting strategy
struct Ascending;
struct Descending;

fn sort_with<T: Ord + Clone, Tag>(_tag: PhantomData<Tag>, data: &mut Vec<T>)
where
    Sorter<Tag>: SortStrategy<T>,
{
    Sorter(PhantomData).sort(data);
}

struct Sorter<Tag>(PhantomData<Tag>);

trait SortStrategy<T> {
    fn sort(&self, data: &mut Vec<T>);
}

impl<T: Ord> SortStrategy<T> for Sorter<Ascending> {
    fn sort(&self, data: &mut Vec<T>) { data.sort(); }
}

impl<T: Ord> SortStrategy<T> for Sorter<Descending> {
    fn sort(&self, data: &mut Vec<T>) { data.sort_by(|a, b| b.cmp(a)); }
}

fn main() {
    let data = &[("name", "Alice"), ("age", "30"), ("city", "Seoul")];

    println!("=== Tag Dispatch: Serialization ===");
    let json = make_serializer::<Json>();
    let xml = make_serializer::<Xml>();
    let csv = make_serializer::<Csv>();

    println!("JSON: {}", json.serialize(data));
    println!("XML:  {}", xml.serialize(data));
    println!("CSV:\n{}", csv.serialize(data));

    // Sorting with tag dispatch
    println!("\n=== Tag Dispatch: Sorting ===");
    let mut nums = vec![3, 1, 4, 1, 5, 9, 2, 6];
    sort_with::<_, Ascending>(PhantomData, &mut nums);
    println!("Ascending:  {:?}", nums);

    sort_with::<_, Descending>(PhantomData, &mut nums);
    println!("Descending: {:?}", nums);
}
